### PR TITLE
Bootstrapping End-to-End Tests

### PR DESCRIPTION
**Pull request by otaviof**
# Changes

This pull-request introduces the machinery to run end-to-end tests, as described on the [enhancement-proposal](https://github.com/shipwright-io/build/blob/main/docs/proposals/cli.md#testing). That means, the following components are ready to use:

- `make verify-kind`: assure KinD's node is ready
- `make install-shipwright`: deploys Tekton and Shipwright Build Controller using the versions specified on the Makefile
- `make test-e2e`: runs the end-to-end tests under `./test/e2e` directory

GitHub Actions have been updated to run [BATs](https://github.com/bats-core/bats-core) e2e tests, in other works, to `make test-e2e` on each pull-request.

# Submitter Checklist

- [x] Includes tests if functionality changed/was added
- [x] ~Includes docs if changes are user-facing~
- [x] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [x] Release notes block has been filled in, or marked NONE

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by otaviof** [comment]
@gabemontero as we've spoken yesterday, here goes the initial scaffolding for end-to-end testing, please consider.

/assign @gabemontero @adambkaplan @coreydaley @HeavyWombat

File: test/e2e/helpers.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+set -eu
+
+BIN="${BIN:-./_output/shp}"
+
+function die () {
+	echo $* >&2
+	exit 1
+}
```
Comment (adambkaplan) [inline test/e2e/helpers.sh:10]: "die" might be considered exclusionary language, in the same category as "kill." I would replace this with "fail"
```suggestion
function fail () {
	echo $* >&2
	exit 1
}
```

File: test/e2e/helpers.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+function shp () {
+	if [ ! -x "${BIN}" ] ; then
+		die "Unable to find '${BIN}' executable"
```
Comment (adambkaplan) [inline test/e2e/helpers.sh:14]:
```suggestion
		fail "Unable to find '${BIN}' executable"
```

**Review by adambkaplan — CHANGES_REQUESTED** [review]
Language change requested, otherwise looks good.

**Comment by gabemontero** [comment]
I take that back  .... https://github.com/shipwright-io/cli/pull/40/checks?check_run_id=3661400810

I'm having a tough time missing items in shipwright PRs today :-/

**Comment by gabemontero** [comment]
One more req @otaviof .... an update to this REPO's readme or some new markdown doc that explains how to add e2e's to the e2e bats file would be good.

WDYT?

**Comment by gabemontero** [comment]
> One more req @otaviof .... an update to this REPO's readme or some new markdown doc that explains how to add e2e's to the e2e bats file would be good.
> 
> WDYT?

in a follow up to https://github.com/shipwright-io/cli/pull/39 I'll take a stab and adding end to end tests for `shp build run <name> --follow`

extra credit will be cancel builds (that one is trickier given the timing windows needed), but we'll see how long it takes to do ^^

**Comment by otaviof** [comment]
> One more req @otaviof .... an update to this REPO's readme or some new markdown doc that explains how to add e2e's to the e2e bats file would be good.
> 
> WDYT?

Good idea! I just pushed the new documentation bits, I'm starting a `docs/testing.md` files to describe the details and show an example of how BATS should be used. Please consider :-)

File: docs/testing.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
+# Testing
+
+All testing tooling is managed by the `Makefile` on this project, therefore all you'll have to do is
+invoking one of those targets to test current changes.
```
Comment (gabemontero) [inline docs/testing.md:4]: invoking -> invoke

File: docs/testing.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+To execute unit-testing present on this project, run:
+
+```sh
+make test-unit
+```
+
+## End-to-End (E2E)
+
+End-to-End tests aim to mimic the real world usage this CLI against the [Shipwright Build
```
Comment (gabemontero) [inline docs/testing.md:22]: usage this -> usage of this

File: docs/testing.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+Controller][shipwrightBuild]. To run end-to-end tests make sure a newly binary is compiled (`make
```
Comment (gabemontero) [inline docs/testing.md:23]: a newly binary is compiled -> you have compiled your latest changes to the binary

File: docs/testing.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+make build test-e2e
+```
+
+### Requirements
+
+The following componets are required to run end-to-end tests.
+
+#### Kubernetes
+
+Before testing a [KinD][kindSig] instance is created, it runs a predefined version of Kubernetes,
```
Comment (gabemontero) [inline docs/testing.md:36]: it runs -> running

put a period instead of a comma Kubernetes

File: docs/testing.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+please consider [GitHub Action files](../.github/workflows/e2e.yaml) for more details. Additionally,
```
Comment (gabemontero) [inline docs/testing.md:37]: capitalize please

File: docs/testing.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+`kubectl` is required to intereact with the Kubernetes Cluster, expected to be installed before
```
Comment (gabemontero) [inline docs/testing.md:38]: expected to be installed -> and you need to install it

File: docs/testing.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+```sh
+make install-shipwright
+```
+
+The install script waits for the Controller instance to be running.
+
+#### BATS
+
+[BATS][batsCore] is a testing framework for Bash. It's structured as a regular script with enhanced
+syntax to define test cases, collect results, and more.
```
Comment (gabemontero) [inline docs/testing.md:55]: Additional sentence:  "You need to install BATS before running the e2e tests".

File: docs/testing.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+### Test Cases
+
+The usual [BATS][batsCore] test-cases can be defined as the following example:
+
+```bash
+@test "short test description, or test name" {
+	# prepare the test-case context running the necessary commands to do so.
+	kubectl ...
+	oc ...
```
Comment (gabemontero) [inline docs/testing.md:71]: remove the oc line .... just cite kubectl for now

**Review by gabemontero — CHANGES_REQUESTED** [review]
Some wording suggestions for the new testing.md doc, but generally looks great @otaviof !

Assuming you agree with my suggestions, I'll lgtm after you have made the updates.

**Comment by otaviof** [comment]
> Some wording suggestions for the new testing.md doc, but generally looks great @otaviof !
> 
> Assuming you agree with my suggestions, I'll lgtm after you have made the updates.

Thanks, @gabemontero! I'm marking all suggestions as resolved, but please consider the latest changes. Appreciated the inputs.

**Review by gabemontero — APPROVED** [review]
/lgtm

**Pull request merged**